syncer/shardddl/optimist_test.go

```diff
@@ -0,0 +1,197 @@
+			[]string{`DROP TABLE bar`}, nil, nil) // both table infos are nil.
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// init with some source tables.
+	err := o.Init(sourceTables)
+	c.Assert(err, IsNil)
+	stm, _, err := optimism.GetAllSourceTables(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(stm, HasLen, 1)
+	c.Assert(stm[task], HasLen, 1)
+	c.Assert(stm[task][source], DeepEquals, o.tables)
+
+	// no info and operation in pending.
+	c.Assert(o.PendingInfo(), IsNil)
+	c.Assert(o.PendingOperation(), IsNil)
+
+	// put shard DDL info.
+	rev1, err := o.PutInfo(info1)
+	c.Assert(err, IsNil)
+	c.Assert(rev1, Greater, int64(0))
+
+	// have info in pending.
+	info1c := o.PendingInfo()
+	c.Assert(info1c, NotNil)
+	c.Assert(*info1c, DeepEquals, info1)
+
+	// put the lock operation.
+	rev2, putted, err := optimism.PutOperation(etcdTestCli, false, op1)
+	c.Assert(err, IsNil)
+	c.Assert(rev2, Greater, rev1)
+	c.Assert(putted, IsTrue)
+
+	// wait for the lock operation.
+	op1c, err := o.GetOperation(ctx, info1, rev1)
+	c.Assert(err, IsNil)
+	c.Assert(op1c, DeepEquals, op1)
+
+	// have operation in pending.
+	op1cc := o.PendingOperation()
+	c.Assert(op1cc, NotNil)
+	c.Assert(*op1cc, DeepEquals, op1)
+
+	// mark the operation as done.
+	c.Assert(o.DoneOperation(op1), IsNil)
+
+	// verify the operation and info.
+	ifm, _, err := optimism.GetAllInfo(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(ifm, HasLen, 1)
+	c.Assert(ifm[task], HasLen, 1)
+	c.Assert(ifm[task][source], HasLen, 1)
+	c.Assert(ifm[task][source][info1.UpSchema], HasLen, 1)
+	c.Assert(ifm[task][source][info1.UpSchema][info1.UpTable], DeepEquals, info1)
+	opc := op1c
+	opc.Done = true
+	opm, _, err := optimism.GetAllOperations(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(opm, HasLen, 1)
+	c.Assert(opm[task], HasLen, 1)
+	c.Assert(opm[task][source], HasLen, 1)
+	c.Assert(opm[task][source][op1.UpSchema], HasLen, 1)
+	c.Assert(opm[task][source][op1.UpSchema][op1.UpTable], DeepEquals, opc)
+
+	// no info and operation in pending now.
+	c.Assert(o.PendingInfo(), IsNil)
+	c.Assert(o.PendingOperation(), IsNil)
+
+	// handle `CREATE TABLE`.
+	rev3, err := o.PutInfoAddTable(infoCreate)
+	c.Assert(err, IsNil)
+	c.Assert(rev3, Greater, rev2)
+	ifm, _, err = optimism.GetAllInfo(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(ifm[task][source][infoCreate.UpSchema][infoCreate.UpTable], DeepEquals, infoCreate)
+	c.Assert(o.tables.Tables[infoCreate.UpSchema], HasKey, infoCreate.UpTable)
+
+	// handle `DROP TABLE`.
+	rev4, err := o.DeleteInfoRemoveTable(infoDrop)
+	c.Assert(err, IsNil)
+	c.Assert(rev4, Greater, rev3)
+	ifm, _, err = optimism.GetAllInfo(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(ifm[task][source][infoDrop.UpSchema], IsNil)
+	c.Assert(o.tables.Tables[infoDrop.UpSchema], IsNil)
+
+	// put another info.
+	rev5, err := o.PutInfo(info2)
+	c.Assert(err, IsNil)
+	c.Assert(o.PendingInfo(), NotNil)
```

Comment:
How about check whether the `PendingInfo` and `PendingOperation` is expected here?

Reply:
added in https://github.com/pingcap/dm/pull/568/commits/6c8c1177b5bdb81164b21e050e690596be3306d4.